### ISSUE TITLE
Positional encoding problem

### ISSUE THREAD

**Issue by entropicemergence** [AkiRusProd/numpy-transformer#1]
Hi, thanks for writing this cupy based transformer, its really help full, but anyway I noticed some probable  error in positional encoding, this part:
x = x + self.pe[:x.shape[0], :]  # (batch_size, seq_len, d_model)

I think it is adding positional encoding toward the batch example, not towards the respective words. That's probably why when I trained your model for more than 10 epoch it the validation loss  would start to increase. At first i think the model start to over fit, but I think this is the problem

**Comment by tingez** [comment]
thanks very much for the numpy based transformer again. 
I also find this possible issue in class PositionalEncoding,
in __init__: self.pe = pe[:, np.newaxis, :].astype(self.data_type)
It seems the shape of self.pe is (max_len, 1, d_model)
In forward, x = x + self.pe[:x.shape[0], :]  # (batch_size, seq_len, d_model)
the shape layout of x is (batch_size, seq_len, d_model),  is not matched with pe (batch_size, 1, d_mode). it seems in this case all the position encoding of input x  will be same as pe[0][0].

It seems 
1. in __init__
self.pe = pe[np.newaxis,:,:].astype(self.data_type) # (1, max_len, d_model)
2. in forward
x = x + self.pe[:, :x.shape[1], :]  # (batch_size, seq_len, d_model)

thanks again

**Comment by AkiRusProd** [comment]
Hi, thanks for your opened issue. I reviewed this problem and amended the positional encoding, the model became better trained, but the problem with overfitting remains not completely resolved. If you have no more proposed amendments, I will close this issue. Thanks.
